### PR TITLE
HPCC-18799 Improve default support for child dataset fields

### PR DESCRIPTION
**Pull request by shamser**
Add support for DEFAULT attribute for child dataset field in record definition. 

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by shamser** [comment]
TODO: create documentation ticket.

**Comment by AttilaVamos** [comment]
@shamser The failed serializetypes problem comes from the master and not related to your PR. About the toxml(..) I don't know,

**Comment by richardkchapman** [comment]
@ghalliday Please review

File: ecl/hql/hqlgram.y [hunk 1]
```diff
@@ -4776,8 +4776,24 @@ fieldAttr
                         {
                             $$.setExpr(createExprAttribute(xmlDefaultAtom, $3.getExpr()), $1);
                         }
-    | DEFAULT '(' constExpression ')'
+    | DEFAULT '(' goodObject ')'
                         {
+                            IHqlExpression * arg = $3.queryExpr();
+                            if (arg->getOperator() == no_temptable && arg->queryChild(0)->queryType()==nullptr)
```
Comment (ghalliday) [inline ecl/hql/hqlgram.y]: Why is this line needed?  At the least it needs a comment.

File: ecl/hql/hqlgram.y [hunk 2]
```diff
@@ -4776,8 +4776,24 @@ fieldAttr
...
+                            {
+                                $3.release().setExpr(createValue(no_null));
+                            }
+                            else
+                            {
+                                ITypeInfo * type = $3.queryExprType();
+                                if (!type || type->getTypeCode() == type_set)
```
Comment (ghalliday) [inline ecl/hql/hqlgram.y]: Does this test add anything to the test in the following lines?  You could use the isNullList() function.

File: ecl/hql/hqlutil.cpp [hunk 3]
```diff
@@ -8855,9 +8859,9 @@ bool ConstantRowCreator::processFieldValue(IHqlExpression * optLhs, ITypeInfo *
         {
             assertex(optLhs);
             IHqlExpression * field = optLhs->queryChild(1);
-            if (!field->hasAttribute(countAtom) && !field->hasAttribute(sizeofAtom))
+            if (!field || (!field->hasAttribute(countAtom) && !field->hasAttribute(sizeofAtom)))
```
Comment (ghalliday) [inline ecl/hql/hqlutil.cpp]: This test doesn't seem right.  I think the problem is that the function is called inconsistently - sometimes it is passed a no_field, and sometimes a no_select.  I suspect it should always be a no_field.
I would add an assert that this was true, and then change processElement() to pass the field instead of the select.

**Review by ghalliday — COMMENTED** [review]
A few comments, and questions.

**Comment by ghalliday** [comment]
@shamser there should also be some other tests which use a non-null dataset.  E.g.,

dataset(r) ds{DEFAULT(DATASET([1,2,3], r))};

since the syntax now supports it.

**Comment by shamser** [comment]
@ghalliday Please can you review the latest commit.

**Comment by AttilaVamos** [comment]
Master is broken. As soon as it will be fixed, I will reschedule this PR.

**Comment by AttilaVamos** [comment]
Master is fixed, this PR is rescheduled.

**Comment by AttilaVamos** [comment]
@shamser Is this childds8 failure related to your changes?

**Comment by shamser** [comment]
@AttilaVamos I'll look into this

**Comment by AttilaVamos** [comment]
@shamser back to you.

**Comment by richardkchapman** [comment]
Also needs rebasing for a clean merge

File: ecl/hql/hqlgram.y (hunk 1 above)
Comment (shamser) [reply ecl/hql/hqlgram.y]: This code is no longer necessary as it is now handled by createTempTableAssign

**Comment by shamser** [comment]
@ghalliday I've fixed the regression.  Please can you review

File: ecl/hql/hqlgram2.cpp [hunk 4]
```diff
@@ -2718,6 +2718,16 @@ void HqlGram::addDatasetField(const attribute &errpos, IIdAtom * name, ITypeInfo
 
     if (queryAttributeInList(virtualAtom, attrs))
         reportError(ERR_BAD_FIELD_ATTR, errpos, "Virtual can only be specified on a scalar field");
+
+    IHqlExpression * defaultAttribute = queryAttributeInList(defaultAtom, attrs);
+    if (defaultAttribute)
+    {
+        IHqlExpression * defaultValue = defaultAttribute->queryChild(0);
+        if (isNullList(defaultValue) || (defaultValue->getOperator() && isNullList(defaultValue->queryChild(0))))
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp]: Should be comparing getOperator() with something (no_temptable?)

File: ecl/hql/hqlutil.cpp [hunk 5]
```diff
@@ -8855,12 +8859,14 @@ bool ConstantRowCreator::processFieldValue(IHqlExpression * optLhs, ITypeInfo *
     case type_groupedtable:
         {
             assertex(optLhs);
+            IHqlExpression * defaultAttrib = optLhs->queryAttribute(defaultAtom);
             IHqlExpression * field = optLhs->queryChild(1);
-            if (!field->hasAttribute(countAtom) && !field->hasAttribute(sizeofAtom))
+            if (defaultAttrib || (!field->hasAttribute(countAtom) && !field->hasAttribute(sizeofAtom)))
```
Comment (ghalliday) [inline ecl/hql/hqlutil.cpp]: not sure why this || is here.  I don't think count or sizeof forms can be generated inline.

File: ecl/hql/hqlutil.cpp [hunk 6]
```diff
@@ -8855,12 +8859,14 @@ bool ConstantRowCreator::processFieldValue(IHqlExpression * optLhs, ITypeInfo *
...
             {
-                if (field->hasAttribute(_linkCounted_Atom))
+                // There may be no field when default value is provided
+                if (field && field->hasAttribute(_linkCounted_Atom))
                 {
-                    if (rhsOp == no_null)
+                    if (rhsOp == no_null || (defaultAttrib && isNullList(defaultAttrib->queryChild(0))))
```
Comment (ghalliday) [inline ecl/hql/hqlutil.cpp]: This code looks out of place - it is likely to build null datasets in the wrong places.

File: ecl/hql/hqlutil.cpp [hunk 7]
```diff
@@ -8891,6 +8890,10 @@ bool ConstantRowCreator::processFieldValue(IHqlExpression * optLhs, ITypeInfo *
                         rtlWriteSize32t(patchPos, out.length() - startOffset);
                         return true;
                     }
+                    else if ( (rhsOp == no_null) || (defaultAttrib && isNullList(defaultAttrib->queryChild(0))) )
```
Comment (ghalliday) [inline ecl/hql/hqlutil.cpp]: This also looks wrong.

File: ecl/hqlcpp/hqltcppc2.cpp [hunk 8]
```diff
@@ -753,7 +753,8 @@ void CChildLinkedDatasetColumnInfo::setColumn(HqlCppTranslator & translator, Bui
     LinkedHqlExpr value = _value;
     ITypeInfo * valueType = value->queryType();
 
-    assertRecordTypesMatch(resultType, valueType);
+    if (value->getOperator() != no_null)
```
Comment (ghalliday) [inline ecl/hqlcpp/hqltcppc2.cpp]: I don't think this should be needed.  Is it still?

File: ecl/hql/hqlgram2.cpp [hunk 9]
```diff
@@ -2718,6 +2718,16 @@ void HqlGram::addDatasetField(const attribute &errpos, IIdAtom * name, ITypeInfo
...
+        {
+            attrs = replaceExpression(attrs, defaultAttribute, createAttribute(defaultAtom,createNullExpr(dsType)));
+        }
+    }
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp]: otherwise this should be checking that the type is compatible.. e.g., if 3 is passed.

**Comment by ghalliday** [comment]
@shamser some of the code doesn't look correct - see comments

**Comment by shamser** [comment]
@ghalliday Please can you have a look at the changes.

**Comment by ghalliday** [comment]
One minor problem.  Adding a default to a field changes the format crc.  I'm not sure if that is a good thing or not.  (Probably good once the field translation code is in use.)
